kore/src/Kore/Attribute/Axiom/Concrete.hs

```diff
@@ -63,27 +74,42 @@ concreteSymbol =
         }
 
 -- | Kore pattern representing the @concrete@ attribute.
-concreteAttribute :: AttributePattern
-concreteAttribute = attributePattern_ concreteSymbol
+concreteAttribute :: [UnifiedVariable Variable] -> AttributePattern
+concreteAttribute = attributePattern concreteSymbol . map attributeVariable
 
 parseConcreteAttribute
-    :: forall variable
-    .  Ord variable
-    => FreeVariables variable
+    :: FreeVariables Variable
     -> AttributePattern
-    -> Concrete variable
-    -> Parser (Concrete variable)
+    -> Concrete Variable
+    -> Parser (Concrete Variable)
 parseConcreteAttribute freeVariables =
-    parseBoolAttributeAux iso concreteId
+    Parser.withApplication concreteId parseApplication
   where
-    iso :: Lens.Iso' (Concrete variable) Bool
-    iso =
-        Lens.iso
-            isConcrete
-            (\b -> Concrete $ if b then freeVariables else mempty)
-
-instance From (Concrete variable) Attributes where
-    from = toBoolAttributesAux (Lens.to isConcrete) concreteAttribute
+    parseApplication
+        :: [Sort]
+        -> [AttributePattern]
+        -> Concrete Variable
+        -> Parser (Concrete Variable)
+    parseApplication params args (Concrete concreteVars) = do
+        Parser.getZeroParams params
+        vars <- mapM getVariable args
+        mapM_ checkFree vars
+        let newVars =
+                if null vars then freeVariables else foldMap freeVariable vars
+        mapM_ (checkDups concreteVars) (getFreeVariables newVars)
+        return (Concrete $ concreteVars <> newVars)
+    checkFree :: UnifiedVariable Variable -> Parser ()
+    checkFree variable =
+        Monad.unless (isFreeVariable variable freeVariables)
+        $ Kore.Error.koreFail
+            ("expected free variable, found " ++ show variable)
+    checkDups :: FreeVariables Variable -> UnifiedVariable Variable -> Parser ()
+    checkDups freeVars var = Monad.when (isFreeVariable var freeVars)
```

Comment:
There are two things about `checkDups` that I think we could do better:

1. It checks for `concrete(X), concrete(X)`, but it doesn't check for `concrete(X, X)`.
1. It doesn't report all the duplicated variables, for example in `concrete(X, Y), concrete(X, Y)` it would report only `X`.